src/devices/src/virtio/block/device.rs

```diff
@@ -258,15 +258,40 @@ impl Block {
                             break;
                         }
                     }
+
                     let status = match request.execute(&mut self.disk, mem) {
                         Ok(l) => {
-                            len = l;
-                            VIRTIO_BLK_S_OK
+                            // Account for the status byte as well.
+                            if let Some(l) = l.checked_add(1) {
+                                len = l;
+                                VIRTIO_BLK_S_OK
+                            } else {
+                                len = 1;
+                                VIRTIO_BLK_S_IOERR
+                            }
                         }
                         Err(e) => {
-                            error!("Failed to execute request: {:?}", e);
                             METRICS.block.invalid_reqs_count.inc();
-                            len = 1; // We need at least 1 byte for the status.
+                            match e {
+                                ExecuteError::Read(GuestMemoryError::PartialBuffer {
+                                    completed,
+                                    expected,
+                                }) => {
+                                    error!(
+                                        "Failed to execute virtio block read request: can only \
+                                        write {} of {} bytes.",
+                                        completed, expected
+                                    );
+                                    METRICS.block.read_bytes.add(completed);
+                                    // Account for the status byte as well.
+                                    len = (completed as u32).checked_add(1).unwrap_or(1);
```

Comment:
`Unwrap()` should be safe here since completed < data len from the request.

Reply:
Hmm, or `unchecked_add()` or we can leave it as it was before: len = completed + 1 and add a comment about why it can not overflow.

Reply:
This is a nit. Please mark it as such.

Reply:
I didn't see it as a nit, since I don't agree to store 1 in used length for this particular issue (overflow). Moreover, if this overflow is never possible, it seemed great to avoid at all the `unwrap_or(1)` and simply use `unwrap()` or the other suggestions. Also, sorry for insisting on this :(.